JBPM-8346: Allow users to select a default filter

nmirasch opened the pull request:
Added the ability to select a saved filter as default

nmirasch commented:
depends on https://github.com/kiegroup/appformer/pull/672
@cristianonicolai @jstastny-cz First approach of the default active filter selection. The option has been added to the saved filters list in a button that for sure have to be reviewed. Please take a look on this, as always, any comment is very welcome. Thanks!

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFiltersPresenter.java:
```diff
@@ -105,4 +113,15 @@ public void onRestoreDefaultFilters() {
     }
+
+    public void onSavedFilterAsDefaultActive(@Observes final SavedFilterAsDefaultActiveEvent event) {
+        YesNoCancelPopup yesNoCancelPopup = YesNoCancelPopup.newYesNoCancelPopup(constants.SavedFilterAsDefaultTitle(),
```
I'm not sure we actually need a confirmation popup in here, just to select a filter as default. It doesnt sound much of a problem if you do it by mistake, wdyt? @jstastny-cz ?

jstastny-cz replied:
I agree in this case it is not necessary, given the user can undo easily.

cristianonicolai commented on jbpm-wb-common/jbpm-wb-data-filters-client/src/main/java/org/jbpm/workbench/df/client/filter/FilterSettingsManagerImpl.java:
```diff
@@ -267,4 +273,37 @@ private FilterSettings createFilterSettings(final String dataSetId,
+    public void defaultActiveFilterInit(final Consumer<FilterSettings> existDefaultActiveFilterCallback,
+                                        final Consumer<FilterSettings> notDefaultActiveFilterCallback) {
+        loadMultiGridPreferencesStore(store -> {
+            if (store.getDefaultGridId() != null && !store.getDefaultGridId().isEmpty()) {
```
```suggestion
            if (store.getDefaultGridId() != null && !store.getDefaultGridId().trim().isEmpty()) {
```

cristianonicolai commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFiltersPresenter.java:
```diff
@@ -55,7 +55,7 @@ public String getTitle() {
     @PostConstruct
     public void init() {
-        loadSavedFilters();
+        filterSettingsManager.getDefaultActiveFilterKey(defaultFilter -> loadSavedFilters(defaultFilter));
```
in here I think default filter is an attribute of the filter itself. You should continue using just _loadSavedFilters_ and that would set the default filter automatically.

cristianonicolai commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFiltersPresenter.java:
```diff
@@ -69,8 +69,12 @@ public void setFilterSettingsManager(final FilterSettingsManager filterSettingsM
+    protected void loadSavedFilters(final String defaultFilter) {
+        view.removeAllSavedFilters();
+        filterSettingsManager.loadSavedFilters(filters -> filters.forEach(f -> {
+            f.setDefaultFilter(f.getKey().equals(defaultFilter));
```
see the previous comment, this probably should be done at the filter manager level for better encapsulation.

cristianonicolai commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/list/AbstractMultiGridPresenter.java:
```diff
@@ -121,11 +122,14 @@ public void init() {
-                              setupActiveSearchFilters();
-                              addDataDisplay(table);
-                          });
+        filterSettingsManager.defaultActiveFilterInit(filterSettings -> addActiveFilters(filterSettings),
```
filter manager should handle the default filter.... if one doesnt exist it should consider creating it, based on the default filter we already apply.

nmirasch commented:
@cristianonicolai @jstastny-cz I updated the code and added some new things:
Now the initial default filter is the one created with the key DEFAULT_FILTER_SETTINGS_KEY at initDefaultFilters. I reviewed the filter to match with the current DefaulfPrototype in each section. The only new one is on task admin and the 'active' status.
This new approach made look strange the first time the user comes to each screen, and that makes me a bit refactoring on the active search filters to maintain, at least the multiselect basic filters checked accordantly with the active filter, removing the clear all basic filters when opening. Please could you take a closer look  on this. In case this could look like a solution to that we can include there more kind of filters to 'process' and match with the ones added at basic filters.
Please take a look and let me know what you think. Thanks in advanced!

cristianonicolai commented:
Jenkins execute full downstream build

cristianonicolai commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFiltersPresenter.java:
```diff
@@ -105,4 +105,11 @@ public void onRestoreDefaultFilters() {
+
+    public void onSaveDefaultActiveFilter(@Observes final SavedFilterAsDefaultActiveEvent event) {
+        filterSettingsManager.saveDefaultActiveFilter(event.getSavedFilter().getKey(), () -> {
+            view.removeAllSavedFilters();
```
reloading all filters seems a bit unnecessary, should just update the icon maybe?

nmirasch replied:
@cristianonicolai I added here a new call updateSavedFiltersDefault to updated the default filter avoiding reload all filters again

cristianonicolai commented on jbpm-wb-common/jbpm-wb-data-filters-client/src/main/java/org/jbpm/workbench/df/client/filter/FilterSettingsManager.java:
```diff
@@ -40,4 +42,8 @@ void getFilterSettings(String key,
         return null;
     }
+
+    void defaultActiveFilterInit(final Consumer<FilterSettings> existDefaultActiveFilterCallback);
```
no need for _final_ keyword in the interface. 
maybe just call it callback?

nmirasch replied:
updated

cristianonicolai commented on jbpm-wb-common/jbpm-wb-data-filters-client/src/main/java/org/jbpm/workbench/df/client/filter/FilterSettingsManagerImpl.java:
```diff
@@ -267,4 +279,40 @@ private FilterSettings createFilterSettings(final String dataSetId,
+    public void defaultActiveFilterInit(final Consumer<FilterSettings> defaultActiveFilterCallback) {
+        loadMultiGridPreferencesStore(store -> {
+            if (store.getDefaultGridId() != null && !store.getDefaultGridId().isEmpty()) {
+                getFilterSettings(store.getDefaultGridId(),
```
no need to break line :)

nmirasch replied:
updated  :)

nmirasch commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/list/AbstractMultiGridPresenter.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
```
@cristianonicolai removed this changes on older files

nmirasch commented on jbpm-wb-executor-service/jbpm-wb-executor-service-client/src/main/java/org/jbpm/workbench/es/client/editors/errorlist/ExecutionErrorListBasicFiltersPresenter.java:
```diff
@@ -103,6 +95,11 @@ public void loadFilters() {
     protected void onActiveFilterAdded(ActiveFilterItem activeFilterItem) {
+        if (activeFilterItem.getKey().equals(constants.Type()) && activeFilterItem.getValue() instanceof List) {
+            final List<String> values = (List<String>) activeFilterItem.getValue();
+            values.forEach(v -> view.checkSelectFilter(constants.Type(),
```
done

nmirasch commented on jbpm-wb-executor-service/jbpm-wb-executor-service-client/src/main/java/org/jbpm/workbench/es/client/editors/requestlist/RequestListPresenter.java:
```diff
@@ -266,25 +273,48 @@ public void setupActiveSearchFilters() {
                                 jobId,
                                 Integer.valueOf(jobId),
-                                v -> removeActiveFilter(equalsTo(COLUMN_STATUS,
+                                v -> removeActiveFilter(equalsTo(COLUMN_ID,
```
done

nmirasch commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFilterView.html:
```diff
@@ -3,6 +3,10 @@
         </button>
+        <button data-field="set-as-default" class="btn btn-link kie-btn-link" type="button" style="color: #252525;"
+                data-toggle="tooltip" data-placement="left" title="...">
+            <lspan data-field="set-as-default-icon" class="fa fa-circle"></lspan>
```
much better, Thanks!

nmirasch commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFilterView.java:
```diff
@@ -47,13 +48,24 @@
+    HTMLButtonElement setDefaultBtn;
+
+    @Inject
+    @DataField("set-as-default-icon")
```
added

nmirasch commented on jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/saved/SavedFilterView.java:
```diff
@@ -47,13 +48,24 @@
+
+    @Inject
+    @DataField("set-as-default-icon")
+    HTMLLabelElement setDefaultIcon;
```
added

nmirasch commented:
@cristianonicolai I was trying to explain the reason because I added the basic filter sync with the saved filter loading at least with the checkboxes. Without this changes, after the default filter loading the user would see a different behaviour than how it's working now, with active filters labels or unchecked checkboxes. And I was guessing if we could extend this to other fields that are added as basic filters like 'Task Name',..

cristianonicolai commented:
@nmirasch I see, but I think it is good as you did here, we don't necessarily need to keep selects in sync.

cristianonicolai commented:
Jenkins execute full downstream build

ederign commented:
@jstastny-cz can you please take a look at this?

nmirasch commented:
@jstastny-cz could you please take a look on this PR? Thanks!

jstastny-cz commented:
Jenkins execute full downstream build

jstastny-cz commented:
@nmirasch I finished the code review, seems fine. Need to see the changes in real before approving, now waiting for the build, I will finish tomorrow.
